sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala

```diff
@@ -21,6 +21,15 @@ import org.apache.spark.annotation.DeveloperApi
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.catalyst.expressions._
 
+// for lazy computing, be sure the generator.terminate() called in the very last
+private[execution] sealed case class LazyIterator(func: () => TraversableOnce[Row])
+  extends Iterator[Row] {
+
+  lazy val results = func().toIterator
+  override def hasNext: Boolean = results.hasNext
+  override def next(): Row = results.next()
+}
```

Comment:
Yeah, sorry I didn't realize the semantics differences between these two iterator classes at first.  I think it's OK to leave `CompletionIterator` as is.